### DISCUSSION TITLE
KFD architect looking into TinyGrad KFD and HSA backends

### DISCUSSION THREAD

**Discussion by fxkamd** [tinygrad/tinygrad#4086] [General]
Long time KFD Architect and Linux kernel hacker here. I'm starting to look into the KFD and HSA backends of TinyGrad to help the ongoing debugging effort. I'm trying to get a top-down view of the situation while others are working from the bottom up looking at GPU firmware, drivers and more. I'm drinking from a fire-hose right now.

I shared my initial observations in a gist: https://gist.github.com/fxkamd/ffd02d66a2863e444ec208ea4f3adc48

I'm still working on finding my way around the code base, how the backends are selected, and how the entry points get invoked in the bigger context. Any helpful pointers are appreciated.

**Comment by geohot** [comment]
Replied to the gist.

Backends are selected with environment variables. KFD=1 will use the KFD backend. DEBUG=4 will show tinygrad debugging info. And IOCTL=1 will show you what ioctls it's making.

**Reply by fxkamd** [reply]
I added another comment on the Gist with some background about the hang and the work we're doing on that.

**Comment by nimlgen** [comment]
@fxkamd, kindly asking if you know of any public documentation on SDMA. I am interested in running indirect buffers (like SDMA_OP_INDIRECT) from SDMA. Is it supposed to work from userspace? What are the fields of SDMA_OP_INDIRECT: priv and csa_addr?
